vowpalwabbit/learner.h

```diff
@@ -678,5 +690,212 @@ void multiline_learn_or_predict(multi_learner& base, multi_ex& examples, const u
+      // function data objects so that if this reduction does not define a function such as
+      // save_load then calling save_load on this object will essentially result in forwarding the
+      // call the next reduction that actually implements it.
+      : common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT, BaseLearnerT>(
+            new learner<DataT, ExampleT>(*reinterpret_cast<learner<DataT, ExampleT>*>(base)), std::move(data), name)
+  {
+    this->_learner->learn_fd.base = make_base(*base);
+    this->_learner->learn_fd.data = this->_learner->learner_data.get();
+    this->_learner->sensitivity_fd.sensitivity_f = (sensitivity_data::fn)recur_sensitivity;
+    this->_learner->finisher_fd.data = this->_learner->learner_data.get();
+    this->_learner->finisher_fd.base = make_base(*base);
+    this->_learner->finisher_fd.func = (func_data::fn)noop;
+
+    set_params_per_weight(1);
+
+    this->_learner->pred_type = base->pred_type;
+    // TODO add label type as something learner knows about itself.
+    // this->_learner.label_type = label_type;
+  }
+
+  reduction_learner_builder<DataT, ExampleT, BaseLearnerT>& set_params_per_weight(size_t params_per_weight)
+  {
+    this->_learner->weights = params_per_weight;
+    this->_learner->increment = this->_learner->learn_fd.base->increment * this->_learner->weights;
+    return *this;
+  }
+
+  learner<DataT, ExampleT>* build() { return this->_learner; }
+};
+
+inline float noop_sensitivity_base(void*, example&) { return 0.; }
+
+template <class DataT, class ExampleT>
+struct base_learner_builder
+    : public common_learner_builder<base_learner_builder<DataT, ExampleT>, DataT, ExampleT, base_learner>
+{
+  base_learner_builder(
+      std::unique_ptr<DataT>&& data, const std::string& name, prediction_type_t pred_type, label_type_t label_type)
+      : common_learner_builder<base_learner_builder<DataT, ExampleT>, DataT, ExampleT, base_learner>(
+            std::move(data), name)
+  {
+    this->_learner->end_pass_fd.func = (func_data::fn)noop;
+    this->_learner->end_examples_fd.func = (func_data::fn)noop;
+    this->_learner->init_fd.func = (func_data::fn)noop;
+    this->_learner->save_load_fd.save_load_f = (save_load_data::fn)noop_sl;
+    this->_learner->finisher_fd.data = this->_learner->learner_data.get();
+    this->_learner->finisher_fd.func = (func_data::fn)noop;
+    this->_learner->sensitivity_fd.sensitivity_f = (sensitivity_data::fn)noop_sensitivity_base;
+    this->_learner->finish_example_fd.data = this->_learner->learner_data.get();
+    this->_learner->finish_example_fd.finish_example_f = (finish_example_data::fn)return_simple_example;
+
+    this->_learner->learn_fd.data = this->_learner->learner_data.get();
+    this->_learner->pred_type = pred_type;
+    // TODO add label type as something learner knows about itself.
+    std::ignore = label_type;
+    // this->_learner.label_type = label_type;
```

Comment:
same here

Reply:
It will come later, probably once all of the reductions are migrated